async-nats/src/nats_error.rs

```diff
@@ -0,0 +1,129 @@
+#![allow(dead_code)]
+
+use std::error::Error;
+use std::fmt::{Debug, Display, Formatter};
+
+// Notice: see unit tests to see how to use this generic error class.
+
+/// The error type for the NATS client, generic by the kind of error.
+pub struct NatsError<Kind>
```

Comment:
I'm not sure about this. The reason is nats already have [the Error type](https://github.com/nats-io/nats.rs/blob/9677baf6dcbf018af16e7f1e295bdddf9d108666/async-nats/src/lib.rs#L150) which is a part of the public interface.

Maybe it is better to use something like `CustomError<T>` here to prevent echoing?

Reply:
Alternatively, I could customize all errors [like this]() as a types of `NatsError`, so that the crate would only provide the `NatsError<T>`, and then make 2 renamings:
- `crate::Error` -> `crate::SourceError` (dynamic)
- `crate::NatsError` -> `crate::Error` (static)

Reply:
The `crate::Error` was added just for having simple alias for typical Rust Boxed error with intention to be used like in all doc examples. It became irrelevant after introducing concrete errors for whole API.

Then it was used for `source`, to have less verbose and consistent syntax at each `source` field. This became irrelevant too, as your improvement gets rid of repeated structs with `source` field.

Changing `crate::Error` name would be a breaking change without good reason. I would prefer to deprecate it.

**Also, this new `NatsError` does not have to be public, as we make all aliases (`type Foo<FooKind>` public).
I was considering if it would "voldemort" the type, but it should be fine.**

Reply:
Oh, and yes, we need to add concrete errors to `service` API. We want the whole API of this crate to be based of concrete errors.

Reply:
Tangentially, crate::Error should in the long term, be removed imo as explicitly saying what it is in say `main` makes for better documentation examples.

This type error type introduced here being public depends on how rustdoc renders it if it was voldemorted.

Reply:
actually, it has to be public, otherwise we would get:

```
error[E0446]: crate-private type `nats_error::NatsError<jetstream::stream::ConsumerErrorKind>` in public interface
    --> async-nats/src/jetstream/stream.rs:1782:1
     |
1782 | pub type ConsumerError = NatsError<ConsumerErrorKind>;
     | ^^^^^^^^^^^^^^^^^^^^^^ can't leak crate-private type
     |
    ::: async-nats/src/nats_error.rs:6:1
     |
6    | pub(crate) struct NatsError<Kind>
     | --------------------------------- `nats_error::NatsError<jetstream::stream::ConsumerErrorKind>` declared as crate-private
```
     
But it does not have to be re-exported as `crate::Error`, but can remain `crate::error::Error`, while deprecating `crate::Error`.

  WDYT @caspervonb ?

Reply:
To be fair, I would not overthink it.
Keep in mind it's an internal API, and often simple and explicit one is better than a smart one. Especially if you want to encourage contributions. The lower the library learning curve, the better :). 

Regarding kinds - this was deliberate decision:
We expose error enum variants only if it makes sense for users to match against them, which is when user can take some action because of error kind returned.

In case of those errors you linked as example it does not make sense for the source kind to exist as separate kind, nor as a kind on returning error level. It's just giving contextual information when logging, which is really easy with those literal &str's.

Reply:
Ahh, I missed your comment yesterday. Anyway, I don't use those new methods `into()` and `with_kind()` so they can be dropped without much sadness ))